Introduce new vars_plugin: recursive_group_vars

Timoses opened the pull request:
##### SUMMARY

This pull request is based on my experience during the integration of Ansible into an existing infrastructure with existing environments with various roles of partly clustered hosts.

It always troubled me to be forced to split up inventories for different environments and use separate group_vars. It conceptually tears apart the infrastructure into smaller infrastructures each with their own sets of variables. But: what if I want to share a variable across environments? What if the separation between environments is not where the separation ends (clusters, host functionalities, ...)? It's not possible to split up the inventory further without serious consequences to the ease of use of Ansible. 

###### Goal

This PR aims at providing a new [vars_plugin](https://docs.ansible.com/ansible/latest/plugins/vars.html) to improve Ansible’s capability of variable assignment by enabling fine-grained control to target specific groups of hosts.

A more detailed description of the current situation follows describing how this PR aims to introduce the new concept of fine-grained control over variable assignment to groups and group combinations through recursive group variables:

###### Parent-Child relation
Ansible works very well when there is a hierarchical structure of parent-child relationships between groups:

![image](https://user-images.githubusercontent.com/6923843/63036461-5d01b480-bebd-11e9-901e-5fc1c5af9d24.png)

In this case, the groups are laid out in a way that the child variables overwrite any parent variables. This means that `webserver-cluster-1` can have different variables than `webserver-cluster-2` by simply writing these into `group_vars/webserver-cluster-1.yaml` and `group_vars/webserver-cluster-2.yaml`.

###### Intersecting groups
Imagine the following scenario:

![image](https://user-images.githubusercontent.com/6923843/63036820-18c2e400-bebe-11e9-9e0d-1c5e42ccd29c.png)

It is no longer possible to set `webserver` as a child of `atlanta` because `webserver` also contains hosts in `boston`. Likewise, `webserver-cluster-1` lists hosts in both `atlanta` and `boston`.

Assuming one would want to assign different variables to all the different clusters in `boston` and `atlanta` it is no longer possible to use parent/child-relationships as a means to prioritize `group_vars`.

Further, it is not possible to use `ansible_group_priority` because `webserver-cluster-X` exists in both `boston` and `atlanta`. Targetting `webserver-cluster-X` would apply to both `atlanta` and `boston` no matter what `ansible_group_priority` is applied.

How to specifically set variables for hosts that exist in `atlanta` and `webserver-cluster-1` and not assign the same variable to the `boston` cluster?

* **The naive solution**: Separate the inventory file into `atlanta/hosts` and `boston/hosts` and use separate group variables (`atlanta/group_vars` and `boston/group_vars`).
  * However, what happens if my `webserver` hosts further separate into `webserver-app` and `webserver-db`? How would I have different variables for e.g. `webserver-cluster-1` -> `webserver-app/db` and `webserver-cluster-2` -> `webserver-app/db`? Again it is not possible to use parent-child relationships since `webserver-app/db` is present in `web-cluster-1` and `web-cluster-2`.
  * Further, splitting up the inventory isolates variables completely. They may not be used in another “isolated” group or environment.
  * It appears that this approach introduces more problems than solutions.

* **Enter recursive group_vars**: What if a `r_group_vars` directory would not only evaluate the first directory-level but traverse further down and match the host's groups along the way? This is what this PR is all about (excerpt from code documentation):
    - In comparison to host_group_vars plugin, this plugin traverses the `r_group_vars` directory
      and instead of matching all files and directories within the first matching directory it
      traverses the directory tree and recursively matches directories to groups during directory
      traversal and applies all files within a matching directory.
    - During traversal in a directory first all files are applied and then matching directories
      traversed. The existing priorities for groups apply during directory traversal.
    - If several matching directories are found on the same directory level they are traversed
      consecutively in a depth-first fashion.
    - The following points are identical to host_group_vars plugin -
    - Files are restricted by extension to one of .yaml, .json, .yml or no extension.
    - Hidden (starting with '.') and backup (ending with '~') files and directories are ignored.
    - Only applies to inventory sources that are existing paths.



With the recursive group_vars plugin enabled the following directory structure is possible:
```
hosts                   # List of hosts and groups
r_group_vars/
    config.yml             # Applies to all clusters anywhere
    atlanta/
       config.yml        # Applies to all hosts in Atlanta
    boston/
        config.yml         # This one is for Boston
        webserver-cluster-1/
            config.yml             # This applies to all webservers of the first cluster in Boston
            webserver-app/
                config.yml         # This only applies to the app webservers of the first cluster in Boston
```


##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
recursive_group_vars

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

The plugin complements the `host_group_vars` and should work together.

- [ ] `version_added` must be adjusted in the `DOCUMENTATION` part of the plugin

##### Open Questions
* How would this plugin integrate with `host_group_vars` plugin? Would there be a priority which is loaded first? Can the user define the order in which they are loaded (i.e. which overwrites the other)?

Commit:
Introduce new vars_plugin: recursive_group_vars

sivel commented on lib/ansible/plugins/vars/recursive_group_vars.py:
```diff
@@ -0,0 +1,95 @@
+        if not isinstance(entities, list):
+            entities = [entities]
+
+        groups = list(filter(lambda ent : isinstance(ent, Group), entities))
```
Please convert this to just a list comprehension:

```suggestion
        groups = [e.name for e in entities if isinstance(e, Group)]
```

Additionally, just convert to a list of names here, since you only ever need that, and not the object later.

sivel commented on lib/ansible/plugins/vars/recursive_group_vars.py:
```diff
@@ -0,0 +1,95 @@
+
+        data = {}
+        if os.path.exists(b_opath):
+            key = '.'.join(list(map(lambda group : group.name, groups)) + [to_text(b_opath)])
```
Use a tuple as a the cache key instead of stringifying this.  This will likely result in a performance problem later if not resolved now.

```suggestion
            key = tuple(groups) + (b_opath,)
```

sivel commented on lib/ansible/plugins/vars/recursive_group_vars.py:
```diff
@@ -0,0 +1,95 @@
+        groups = list(filter(lambda ent : isinstance(ent, Group), entities))
+
+        if not groups:
+            return dict()
```
Use dict literals:

```suggestion
            return {}
```

sivel commented on lib/ansible/plugins/vars/recursive_group_vars.py:
```diff
@@ -0,0 +1,95 @@
+from ansible.inventory.group import Group
+from ansible.utils.vars import combine_vars
+
+FOUND = {}
```
Does this really need to be a global?  Can it not be a property on the `VarsModule` instead>

Timoses replied:
I started out with the `host_group_vars.py` and used most of it. `FOUND` is global there as well.

I tested it and adding it to the class scope appears to work as well.
```
class VarsModule(BaseVarsPlugin):
    FOUND = {}
```
Acts a bit like a static variable I suppose (I'm new to Python).

Timoses replied:
Committed with `FOUND` in class scope.

sivel commented on lib/ansible/plugins/vars/recursive_group_vars.py:
```diff
@@ -0,0 +1,95 @@
+        # Apply matching files in path
+        for group in groups:
+            try:
+                found_files = loader.find_vars_files(path, group.name, allow_dir=False)
```
With `groups` below changing to just be a list of strings, this should be changed

```suggestion
                found_files = loader.find_vars_files(path, group, allow_dir=False)
```

sivel commented on lib/ansible/plugins/vars/recursive_group_vars.py:
```diff
@@ -0,0 +1,95 @@
+            return dict()
+
+        super(VarsModule, self).get_vars(loader, path, entities)
+        b_opath = os.path.realpath(to_bytes(os.path.join(self._basedir, 'r_group_vars')))
```
The `to_bytes` should be on the `self._basedir` instead:

```suggestion
        b_opath = os.path.realpath(os.path.join(to_bytes(self._basedir), b'r_group_vars'))
```

sivel commented on lib/ansible/plugins/vars/recursive_group_vars.py:
```diff
@@ -0,0 +1,95 @@
+            try:
+                found_files = loader.find_vars_files(path, group.name, allow_dir=False)
+                for found in found_files:
+                    self._display.debug("recursive_group_vars - Matched file : " + to_text(found))
```
Try using `.format()` or `%s` formatting instead:

```suggestion
                    self._display.debug("recursive_group_vars - Matched file : %s" % to_text(found))
```

sivel commented on lib/ansible/plugins/vars/recursive_group_vars.py:
```diff
@@ -0,0 +1,95 @@
+    def find_vars_recurse(self, loader, groups, path):
+        data = {}
+
+        self._display.debug("recursive_group_vars - Traversing dir : " + path + " with groups : " + to_text(groups))
```
```suggestion
        self._display.debug("recursive_group_vars - Traversing dir : %s with groups : %s" % (path, to_text(groups)))
```

sivel commented on lib/ansible/plugins/vars/recursive_group_vars.py:
```diff
@@ -0,0 +1,95 @@
+                    if new_data:
+                        data = combine_vars(data, new_data)
+            except Exception as e:
+                raise AnsibleParserError(to_native(e))
```
We should probably do more here than just reraise the exception with a different type. Add a message here so it gives context about the failure.  The same with the next one too.

sivel commented on lib/ansible/plugins/vars/recursive_group_vars.py:
```diff
@@ -0,0 +1,95 @@
+        # Recurse into directories
+        for group in groups:
+            try:
+                b_opath = os.path.realpath(to_bytes(os.path.join(path, group.name)))
```
```suggestion
                b_opath = os.path.realpath(to_bytes(os.path.join(path, group)))
```

sivel reviewed (changes requested):
This is largely just a code review

However, on a bikeshedding note, I'm not fond of the `r_group_vars` name.  It's not very explicit and self explanatory.  I wonder if we couldn't come up with something similar to `ansible_group_priority` that could be used within a normal group vars file instead.

Timoses commented:
> This is largely just a code review

Thanks for the review!

> However, on a bikeshedding note, I'm not fond of the `r_group_vars` name. It's not very explicit and self explanatory.

I am not sure whether there could be a better name other than `r_group_vars`. Perhaps write it out to `recursive_group_vars`? Maybe somebody has a better idea for a name.

>  I wonder if we couldn't come up with something similar to `ansible_group_priority` that could be used within a normal group vars file instead.

The problem is that one would have to be able to assign variables for specific combinations of inventory groups. If you'd like this in the inventory then it could look something like the following:
```
[atlanta:webserver-cluster-1:vars]
; Variables that only apply to cluster 1 in atlanta
xyz = abc
```
In the `atlanta:webserver-cluster-1` part instead of a colon `:` an ampersand `&` could be used as well.. (This is freely invented syntax for illustrating a possible design).
However, I do not see how this would be better than an `r_group_vars` directory which comes with a nice tree overview of the group combinations (the inventory file variant would only be a plain file which is more difficult to navigate and make sense of).

Timoses commented:
bot_status

sivel commented:
Just to provide at least 1 further update here, this PR will at minimum be delayed until we add new functionality to whitelist vars plugins. This is one of the first community contributed vars plugin we have had, and it has raised the concern that we need a facility like that of callbacks, where additional vars plugins must be enabled.

Our team is looking into writing this functionality.

Timoses commented:
> Just to provide at least 1 further update here, this PR will at minimum be delayed until we add new functionality to whitelist vars plugins. This is one of the first community contributed vars plugin we have had, and it has raised the concern that we need a facility like that of callbacks, where additional vars plugins must be enabled.
> 
> Our team is looking into writing this functionality.

That sounds very good. Since this plugin may simply use another directory (`r_group_vars`) both vars plugins can work complementary. As such, it could be worth looking into the possibility to set which is loaded first (i.e. the user can set in which order the plugins are loaded to influence which has priority and overwrites the variables from the other).

Would there be value in taking this into a [core meeting](https://github.com/ansible/community/tree/master/meetings#core-team-meeting)?

sivel commented:
> Would there be value in taking this into a [core meeting](https://github.com/ansible/community/tree/master/meetings#core-team-meeting)?

absolutely

Timoses commented:
Core meeting subject suggested: https://github.com/ansible/community/issues/482#issuecomment-522209527

willthames commented:
We manage to cope with this situation by using combinations of groups so that you could easily have a group called webserver-cluster-boston that overrides variables set in webserver-cluster or set in boston

For large inventories, I suggest using the generator inventory plugin: http://willthames.github.io/2017/11/01/generating-inventory.html

Commit:
Fix sanity check docs-build

Timoses commented:
> We manage to cope with this situation by using combinations of groups so that you could easily have a group called webserver-cluster-boston that overrides variables set in webserver-cluster or set in boston
> 
> For large inventories, I suggest using the generator inventory plugin: http://willthames.github.io/2017/11/01/generating-inventory.html

Thanks for the suggestion. We do use our own script for generating the inventory file(s).

Creating extra groups for the purpose of combining other groups sounds like a nice work-around. However, a work-around means that there's a lack conceptually within Ansible's way of mapping variables to groups (or combinations thereof). 
With many varying groups in different depths this can quickly become messy. For one thing, how do you create rules on how the combinations of groups are created: Is it `boston-webserver-1-webserver-app` or `webserver-app-webserver-1-boston`? One would have to potentially generate all possible combinations in order to support them in the `group_vars` directory. Aside from being quite ugly, it would also make the `group_vars` directory (and inventory file) very cluttered and harder to organize.

I believe the `r_group_vars` approach is very clean and lean as directories only have to be created under `r_group_vars` directory when the user wishes to target specific hosts within groups (e.g. `r_group_vars/boston/webserver-1/webserver-app.yml`). It also offers flexibility to the user (`r_group_vars/webserver-app/webserver-1/boston.yml` works just as well; of course, group priorities have to be considered as well).

willthames commented:
> I believe the r_group_vars approach is very clean and lean as directories only have to be created under r_group_vars directory when the user wishes to target specific hosts within groups (e.g. r_group_vars/boston/webserver-1/webserver-app.yml). It also offers flexibility to the user (r_group_vars/webserver-app/webserver-1/boston.yml works just as well; of course, group priorities have to be considered as well). The inventory remains logically organized.

All groups are created by the generator plugin, in whatever order you choose:
```
hosts:
    name: "{{ operation }}-{{ application }}-{{ environment }}-runner"
```
so here the order would always be operation, application then environment. But it's designed to be flexible so that you can choose the order that makes the most sense

We only create directories (or occasionally files) under group_vars when we actually need them - we definitely don't need to create every single one. So yes, it's getting a little cluttered as we grow but it's still pretty easy to find things - and furthermore, it's obvious where to look.

One downside of your approach is what happens if both `r_group_vars/boston/webserver-1/webserver-app.yml` and `r_group_vars/webserver-app/webserver-1/boston.yml` exist? Which one wins?

I'm not really against this idea, I just want to ensure that people are informed that alternative problems to this solution exist, and that there are trade offs.

Commit:
Improve documentation

Timoses commented:
> All groups are created by the generator plugin, in whatever order you choose:
> 
> ```
> hosts:
>     name: "{{ operation }}-{{ application }}-{{ environment }}-runner"
> ```
> 

That makes sense and I think this is a vaiable alternative solution. IMO the downside is that the inventory will contain groups that can be easily created by intersection of other groups and thus feels unnecessary.

---

> We only create directories (or occasionally files) under group_vars when we actually need them - 

You'd still have all assignments in one directory (and you have to know what the group names represent; which of course you do when you design the inventory groups and set a predefined order following a naming scheme for combinations of groups).

---

> 
> One downside of your approach is what happens if both `r_group_vars/boston/webserver-1/webserver-app.yml` and `r_group_vars/webserver-app/webserver-1/boston.yml` exist? Which one wins?

There's no downside. It is described in the documentation of the plugin.

E.g. if the following is true regarding `ansible_group_priority`:
```
boston < webserver-{number} < webserver-app
```

Then in the following scenario:
```
r_group_vars/
    boston/
        webserver-1/webserver-app.yml
        webserver-app.yml
    webserver-1/
        webserver-app.yml
```

In the `boston` directory: `webserver-1/webserver-app.yml` will win over `webserver-app.yml` (files are matched and applied first -> then sub-directories traversed and evaluated in equal fashion).

The interesting part in that scenario is that `r_group_vars/webserver-1/webserver-app.yml` would be prioritized over anything in `boston` (due to `ansible_group_priority` or naturally the alphabetic order if no priority were defined).
This goes against the "The deepest found matching file is applied last." in the documentation -> This should be clarified/discussed: What is the desired behaviour?
1. Apply most specialized (i.e. deepest) last? In above scenario it would mean that `boston -> webserver-1 -> webserver-app.yml` would win over `webserver-1 -> webserver-app`.
1. Keep it simple (as it is currently implemented): First apply files in a directory and then traverse directories in the usual order: https://github.com/ansible/ansible/blob/100b56439ee0fb779f049c1cd61ebb305a6bb5ce/lib/ansible/inventory/helpers.py#L26

I would vote for the second option: It makes reasoning about priorities much easier. Option 1 would mean: In one directory level I would never be sure whether the priority is followed as I would need to inspect all sub-directories and find the "longest path".

-> I will remove the "The deepest found matching file is applied last." from the documentation.
Please provide any thoughts if you think option 1 is preferable.

---

> I'm not really against this idea, I just want to ensure that people are informed that alternative problems to this solution exist, and that there are trade offs.

I don't see any tradeoffs in using `r_group_vars`. If you set a naming scheme:
```
"{{ operation }}-{{ application }}-{{ environment }}-runner"
```

Then you'd follow the following `r_group_vars` structure:
```
r_group_vars/
    operation-X/
        application-Y/
            environment-Z.yml
```

Timoses commented:
ready_for_review

Commit:
Load all files in matching directories

Timoses commented:
I have had a change of heart:
Loading only files that match a group of the current host turns out to be rather limiting.

Instead, now the plugin will traverse the directory tree and load all files that are contained within a directory path that matches the host's groups.
Shallower files are applied first, deeper files last. This allows specialization.

Timoses commented:
Rebased to current `devel` branch.

Timoses commented:
@sivel : 
* Rebased to current `devel` branch.
* Adjusted PR initial message to contain updates from Commit "Load all files in matching directories"

Not sure what the 'Unstable' CI check is about.

grunzi commented:
I would love to see this feature getting accepted. Really good work.

bcoca commented:
Hi!

Thank you very much for your submission to Ansible. It sincerely means a lot to us that you've taken the time to contribute.

Per recent announcements at AnsibleFest Atlanta 2019, going forward we are no longer accepting new one off modules/plugins, or modules/plugins that create a new category.

However, we recommend looking into providing this functionality through Ansible Galaxy via Ansible Collections. You can find more information about collections at https://docs.ansible.co
m/ansible/devel/dev_guide/developing_collections.html.

Thank you once again for this and your interest in Ansible!

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project
